cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/AppMetadataStore.java

```diff
@@ -1240,11 +1250,141 @@ public long getProgramRunCount(ProgramId programId) {
     return result;
   }
 
+  void createStartTimeIfNotExist() {
+    MDSKey mdsKey = new MDSKey(RUN_COUNT_FIRST_UPGRADE_TIME);
+    if (get(mdsKey, Long.TYPE) == null) {
+      LOG.info("Start to upgrade the run count");
+      write(mdsKey, TimeUnit.MILLISECONDS.toSeconds(System.currentTimeMillis()));
+    }
+  }
+
+  /**
+   * Compute the old run count if the run record is older than first start up time after 5.1
+   */
+  boolean computeOldRunCount(int maxRows) {
+    boolean upgradeDone = upgradeActiveRowkey(TYPE_RUN_RECORD_STARTING, maxRows);
+    upgradeDone &= upgradeActiveRowkey(TYPE_RUN_RECORD_STARTED, maxRows);
+    upgradeDone &= upgradeActiveRowkey(TYPE_RUN_RECORD_SUSPENDED, maxRows);
+    upgradeDone &= computeOldCompletedRunCount(maxRows);
+    return upgradeDone;
+  }
+
+  /**
+   * Upgrades the rowkeys for the active run record. Note that we do not count the run record of the old active runs
+   * since it is hard to deal with the races and we expect user to stop all the programs before they do the upgrade.
+   *
+   * @param recordType type of the record
+   * @param maxRows    maximum number of rows to be upgraded in this call.
+   * @return true if no rows required an upgrade
+   */
+  private boolean upgradeActiveRowkey(String recordType, int maxRows) {
+    LOG.info("Checking upgrade for {}", recordType);
+    MDSKey startKey = new MDSKey.Builder().add(recordType).build();
+    Map<MDSKey, RunRecordMeta> oldMap = listKV(startKey, RunRecordMeta.class, maxRows);
+
+    for (Map.Entry<MDSKey, RunRecordMeta> oldEntry : oldMap.entrySet()) {
+      MDSKey oldKey = oldEntry.getKey();
+      RunRecordMeta runRecord = oldEntry.getValue();
+      // extract the program id from the old key, the format will be same after the program name is reached, so the
+      // method can be used directly.
+      MDSKey newKey = getProgramKeyBuilder(TYPE_RUN_RECORD_ACTIVE, getProgramID(oldKey))
+        .add(runRecord.getStartTs())
```

Comment:
the timestamp needs to be inverted

Reply:
also, please update the javadoc at the top of this class with the new rowkey format

Reply:
The new key format is already in the javadoc of this class